File: src/main/java/br/com/hu/controller/CurrencyController.java
package br.com.hu.controller;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import br.com.hu.service.CurrencyService;

@RestController
@RequestMapping("v1")
public class CurrencyController {

	@Autowired
	private CurrencyService service;

	@RequestMapping(value = "/currency", method = RequestMethod.GET)
	public ResponseEntity<?> converter(@RequestParam String from, @RequestParam String to,
			@RequestParam String amount) {
		try {
			Double conversion = service.conversion(to, from, amount);
			return new ResponseEntity<Double>(conversion, HttpStatus.OK);
		} catch (NumberFormatException e) {
			return new ResponseEntity<String>("The amount format is wrong", HttpStatus.BAD_REQUEST);
		} catch (NullPointerException e) {
			return new ResponseEntity<String>("Paramter can not be null", HttpStatus.BAD_REQUEST);
		} catch (Exception e) {
			return new ResponseEntity<String>(e.getMessage(), HttpStatus.BAD_REQUEST);
		}
	}
}

File: src/main/kotlin/quoteservice/web/controllers/CurrencyController.kt
package quoteservice.web.controllers

import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.*
import quoteservice.loggers.Logger
import quoteservice.web.entities.requests.AddCurrencyRequest
import quoteservice.web.entities.responses.ConversionResponse
import quoteservice.web.services.RateService
import quoteservice.web.suppliers.RatesSupplier

@Controller
@RequestMapping("/currency")
class CurrencyController (
    @Qualifier("forexSupplier") val forexSupplier: RatesSupplier,
    val rateService: RateService
) {

    @GetMapping("/convert")
    fun convert(
        @RequestParam from: String,
        @RequestParam to: String,
        @RequestParam amount: String
    ): ResponseEntity<ConversionResponse> {
        val result = rateService.convert(baseCurrencySymbol = from, targetCurrencySymbol = to, amount = amount)
        return ResponseEntity.ok(result)
    }

    @PostMapping("/add")
    fun add(@RequestBody request: AddCurrencyRequest): ResponseEntity<String> {
        rateService.add(request)
        return ResponseEntity(HttpStatus.NO_CONTENT)
    }

    @DeleteMapping("/delete")
    fun delete(@RequestParam currency: String): ResponseEntity<String> {
        rateService.delete(currency)
        return ResponseEntity(HttpStatus.NO_CONTENT)
    }
}
